### ISSUE TITLE
Ability to print a quote or estimate.

### ISSUE THREAD

**Issue by SteveIreland** [jekkos/opensourcepos#1061]
## Enhancement

I'm pretty sure I stumbled across this request mentioned on another issue but I can't find it.   One of the features I need to add soon after the kit enhancements is the ability to print a quote.

From what I can tell I'll need to work off of the suspended sales tables.  I think that was mentioned as a possible solution and that works perfectly fine for me.

Is anyone working on this now or is that a contribution that would be desired for this project or should I just add it to my fork?

**Comment by odiea** [comment]
#352 works for me.

**Comment by daN4cat** [comment]
@SteveIreland welcome to add this if you find the time.

**Comment by SteveIreland** [comment]
I had hoped to include it as part of the item-kit branch but, as you probably anticipated, the item-kit changes were a bigger bite than I expected so I didn't have time to incorporate it yet.  I have to admit there was a lot more to learn than I thought and I still have more to learn.

I'll take a look at #352 and put together something for it - it looks like I'll have to brush up on my Spanish to absorb all of the ticket but I'll get there ... eventually.

**Comment by SteveIreland** [comment]
I've started putting together some documentation on the Wiki about how I envision this coming together.

I particularly want to draw the attention of those that are interesting to the section on configuration found at [Sales.Invoice.Configuration](https://github.com/jekkos/opensourcepos/wiki/Sales#4-configuration).

Feedback is always appreciated.

**Comment by jekkos** [comment]
@SteveIreland feel free to update the Mysql model any time. Regarding the invoice format: it would be nice to use an object oriented design there. 

In this case I mean that we should have Variable class instances in a list that have parsing and repalce logic inside. Then an abstraction can be used from which the correct instance decides whether it has to replace the text with a real value.

**Comment by SteveIreland** [comment]
@jekkos oh my!  My specialty is "quick and dirty procedural code", but for you I will climb that mountain!

Seriously though, my full time (i.e.paid) job is JEE Java Web development so object based development is not unfamiliar to me, but my PHP growth stalled when I stopped teaching PHP development at a community college over a decade ago.  So I have a bit of a learning curve to go through ... please be patient.

**Comment by jekkos** [comment]
@SteveIreland good to hear that. I just meant that this code is procedural and rather cumbersome at this moment. Just a small suggestion in case you were planning to extend it.

I've been doing Java as well for some time but have now switched to scala which is were the heap bytes go to die..

**Comment by SteveIreland** [comment]
I'm sorry to admit that scala is a language I wouldn't recognize if I saw it.  Another book to add to my languages library!

**Comment by daN4cat** [comment]
@SteveIreland Scala is strong in Data Science / Big Data sector. With that comes a good salary. At least that's the situation in London. I don't personally deal with coding anymore outside this petty project.

**Comment by daN4cat** [comment]
And it builds on top of Java, so good to learn for you

**Comment by SteveIreland** [comment]
I'll have to pass on Scala as a career opportunity - but thanks for the suggestion.  I am simply too comfortable with where I am to start another programming language.

As for this issue I've already started coding it up and must say that I am impressed with the flexibility of the design of OSPOS.  I've managed to seamlessly integrate invoicing and quotes into the register UI as if it's been there all along.  I've simply added them as a couple of additional Register Mode options.  I made this change in direction while I was watching the Super Bowl - sorta like the score.

I haven't done it yet, but I will update the wiki this week to describe where I'm headed with it.

**Comment by SteveIreland** [comment]
One question for the community is this.  My client is going to have one person using OSPOS primarily as register - so I can see where that employ might need to be "locked in" to a particular register mode.  Then other employees might be salespersons that do quotes and invoicing.  Is anyone else interested in being able to lock an employee into a particular register mode?

**Comment by odiea** [comment]
I think if you add it as an app config item It would give the admins or managers the ability to select if an Employee can see just their transactions.

**Comment by SteveIreland** [comment]
That's a slightly different issue but I can include that in this set of changes and if I do that then that will require a change to the employee table and if I do that I might as well add support for the other.  One field would support restricting an employee to their transactions and the other would delineate the register modes that are valid for the employee.

**Comment by odiea** [comment]
I did it pretty rudimentary and just added the following into the temp tables in Models Sale.php so Takings would show just the Sales for the current employee.
if (!empty($this->config->item('current_employee_only')))
		{	
		$employee_id = $this->Employee->get_logged_in_employee_info()->person_id;
		$this->db->where('employee_id', $employee_id);
		}

**Comment by SteveIreland** [comment]
Yes, that's pretty much the code I would steal/add except that I wouldn't want all employees restricted - just selected ones. Which configuration panel did you add "current_employee_only" to?

**Comment by jekkos** [comment]
see also #204 #382 .. a couple of requests in the past for a more fine grained permission model

**Comment by odiea** [comment]
I added it to general Config. My thought was only the logged in user can see their info. I assume the program will be on a main server and all others login from a different computer only seeing their work.

**Comment by odiea** [comment]
This does not effect the Detailed Sales Report so I figured anyone who had access to that would be able to see what's going on with other terminals.

**Comment by odiea** [comment]
Yes I see your point. As I stated in a similar issue What is an employee to do if they need to reprint a receipt from a previously logged on employ (Who's out to lunch) and the Manager is missing?

**Comment by odiea** [comment]
I have always used the all or none theory. I think it would work better as an Employee permission. Then others would not be locked out  which is the case using what I have.

**Comment by odiea** [comment]
I changed my code to work with already added permissions. Just like Takings.

if (!$this->Employee->has_grant('reports_sales', $this->session->userdata('person_id')))
		{
		$employee_id = $this->Employee->get_logged_in_employee_info()->person_id;
		$this->db->where('employee_id', $employee_id);
		}

**Comment by SteveIreland** [comment]
@odiea I have always been accused of adding so many configuration options that it is overwhelming.  So I'll need you to keep me "in check".  :-)

I'm going to go ahead and finish out the work on the invoice and quote changes and hopefully publish that for review this weekend.  Then while waiting on feedback I'll review the employee changes and the other two issues that @jekkos mentioned.

**Comment by jekkos** [comment]
In fact adding permissions shouldn't be too hard. Probably two checks should be added. One to hide specific controls in the view, and another to avoid the resulting action taking place. After that it's just a matter of adding a permission and grant to the db scripts

**Comment by jekkos** [comment]
One could argue that a condition in the view is good enough but even then a user with some technical background can still fire the http requests manually

**Comment by SteveIreland** [comment]
Believe it or not, I am actually on target for pushing out the quote and invoice changes this weekend for review and feedback.   I've updated the Wiki with a few of the design decisions made while coding it up.  

When I prepare the pull request I'll make a note of the things that I think would be of interest to those that want to consider using it.  My goal is to have zero impact on those that don't want to use it (other than making sure that the configuration is set to NOT use it which should be the default).

**Comment by SteveIreland** [comment]
The pull request has been pushed. 

To engage the quotes and invoices features Invoice needs to be enabled in the configuration.  This adds two new register modes ... "Sales by Invoice" and "Quote".  You should freely be able to toggle between all register modes at any time.

The new tokens (i.e. {ISEQ} and {QSEQ} are supported.  The old $CO and $SCO are not yet supported but have an equivalent in {CO} and {SCO}.  The fixed length for a token is supported (i.e. {ISEQ:6} )

A quote is automatically suspended when it is printed and can be retrieved and reprinted with the same quote number.

**Comment by daN4cat** [comment]
@SteveIreland thank you for the contribution:

- I reviewed the code and added my comments.
- I hope this feature solves the pro-forma invoice issue that was discussed many times in the past. I made this issue an Epic and linked all the old pro-forma like issues to it. Maybe you want to read the use cases so we get one solution catching all forever

General thought, @jekkos to comment please:

- I sense that the whole invoicing in all the variants / use cases is growing. To keep the sale code clean we should consider isolation it out in a Invoice_lib or inside Sale_lib. Maybe Invoice_lib could do?
- I sense that we are increasing the possible last step after the sale transaction is ended. We can print, we can email, we will sms a message, it's a receipt (many templates), it's an invoice, it's a quote...
all of this needs a thought and make the sale & sale register something generic with actions attached.
It's probably not a 1 day job but something to start considering.

**Comment by SteveIreland** [comment]
I know I'm pushing my luck but if this looks pretty close to what's needed then can we go ahead and merge it in and I'll spend the the coming week addressing all of the remaining issues (at least those that are within my skill set).

**Comment by daN4cat** [comment]
I leave to @jekkos to merge this as I don't use invoicing and I did little test of this in the past just to amend things.

**Comment by jekkos** [comment]
Yes indeed I guess we need to take action and reduce the amount of code there. Some of the methods in Sales controller have become too tedious. 

I think that it would make sense to have some kind of hook in the end on which we can register a generic sort of action fires. This could be cleanly done in an OO way by providing a simple interface method and extracting the specific functionalities accordingly.

**Comment by SteveIreland** [comment]
@jekkos @odiea @daN4cat  

Since I just added the ability to generate an unpaid or partially paid invoice I will now need to add the ability to add payments to a sales of type invoice.  That means I need to be able to add payments to a sale.

I need to a payment option to on the takings update to capture payment type, payment date, and payment amount.  These fields option will only be visible if the invoice has a balance due.

**Comment by daN4cat** [comment]
Have you checked the sale edit box? I added logic to edit payment type and if one payment is 0 it removes it automatically (it's a known issue that people tend to add 0 payment type by mistake and then you can remove later, with this edit it's gleaned automatically).
In case of partial payment you can extend to add payment to balance. Please make sure payments can not be removed otherwise we allow tax frauds cases or dodgy sale assistants remove cash payments to hide cash, which we don't want.

This feature of partial payments was requested several times. It's better to cover it under a dedicated topic and link past requests to it. I'll do that this evening.

**Comment by odiea** [comment]
In the current version with just changing 1 line of code you can make payments and they do show just fine in Takings. I think the biggest issue would be to create an additional module to be able to make additional payments and Reports .
![2017-02-14 05_29_21-wedidit _ powered by ospos 3 1 0](https://cloud.githubusercontent.com/assets/10962177/22929027/aa270438-f276-11e6-813d-d7d0f5ed7488.png)

**Comment by alvin101** [comment]
This is another enhancement I think best left for a back office accounting or erp sys...

**Comment by SteveIreland** [comment]
I started to get all indignant and and question @alvin101 regarding his bias for Phreebooks. 

But after the second shot of whiskey I realized how "on target" the comment was.  I swear I heard a choir of angels (or maybe it was the background music for 2001 A Space Odyssey - I can't be sure).

Rethinking again.  Thanks @alvin101

**Comment by daN4cat** [comment]
All this typing from my side and just one single sentence did it all.... better to watch TV next time 😊

**Comment by daN4cat** [comment]
@SteveIreland can this be closed now?

**Comment by SteveIreland** [comment]
This definitely can be closed.   Thanks.
